### PR TITLE
API to perform PCA using SVD.

### PR DESCRIPTION
**Pull request by vruano**

### CODE REVIEW

**Comment by vruano** [comment]
@samuelklee please take a look.

File: src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ */
+public final class PCA {
+
+    /**
+     * Array of variable centers subtracted from the data-matrix before PCA.
+     */
+    private final double[] centers;
+
+    /**
+     * Principal component loadings (i.e. eigen-vectors).
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java]: "eigen-vectors -> eigenvectors"

also, i'm not sure about the terminology, but are "loadings" typically orthonormal?  i think they are usually scaled by the square roots of the eigenvalues.  maybe just call this field "eigenvectors" and note their normalization in the doc.

File: src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ * Unit tests for {@link PCA}.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class PCAUnitTest {
+
+    /**
+     * Precision required when comparing eigen vector values.
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java]: "eigen vector" -> "eigenvector"

File: src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+import java.util.Arrays;
+import java.util.stream.DoubleStream;
+
+/**
+ * Unit tests for {@link PCA}.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class PCAUnitTest {
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java]: test class can be final (although we are inconsistent about this)

File: src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            Assert.assertEquals(v1.getEntry(i), v2.getEntry(i), epsilon, " Differences between " + Arrays.toString(v1.toArray()) + " and " + Arrays.toString(v2.toArray()));
+        }
+    }
+
+    private void assertEqualLoadings(final RealMatrix m1, final RealMatrix m2, final RealVector variances, final double epsilon) {
+        Assert.assertEquals(m1.getRowDimension(), m2.getRowDimension());
+        Assert.assertEquals(m1.getColumnDimension(), m2.getColumnDimension());
+        final boolean[] sameSign = new boolean[m1.getColumnDimension()];
+
+        // Find out whether the eigen vector is the same or inverted (which is equivalent).
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java]: eigenvector

File: src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        final JavaSparkContext sparkContext = SparkContextFactory.getSparkContext("test", SparkContextFactory.DEFAULT_SPARK_MASTER);
+        final PCA pca = PCA.createPCA(dataMatrix, dm -> SVDFactory.createSVD(dm, sparkContext));
+        final RealVector centers = pca.getCenters();
+        final RealVector variances = pca.getVariances();
+        final RealMatrix loadings = pca.getLoadings();
+        Assert.assertNotNull(loadings);
+        assertEquals(centers, new ArrayRealVector(TEST_EXPECTED_CENTERS), EPSILON);
+        assertEquals(variances, new ArrayRealVector(TEST_EXPECTED_VARS), EPSILON);
+        assertEqualLoadings(loadings, new Array2DRowRealMatrix(TEST_EXPECTED_LOADINGS), variances, EPSILON);
+
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java]: whitespace

**Comment by samuelklee** [comment]
looks fine to me!  just a few very minor things.  back to @vruano

File: src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java (hunk 1 above)
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java]: Yup, you are right.

File: src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java (hunk 3 above)
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/utils/pca/PCAUnitTest.java]: Same.

**Comment by vruano** [comment]
Back to @samuelklee

**Comment by samuelklee** [comment]
i think you missed replacing some uses of "loadings" in both classes.  OK to merge after these are fixed (be sure to squash, as @vdauwera reminded me over in gsa-unstable)

**Comment by vruano** [comment]
sure.

**Comment by vruano** [comment]
Test seems to pass locally, merging.

**Pull request merged**